### PR TITLE
Refactor HMR hooks

### PR DESCRIPTION
**Pull request by ooflorent**
Refactor HMR plugin to change how hooks are installed.

**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

already tested

**Does this PR introduce a breaking change?**

yes

**What needs to be documented once your changes are merged?**

Currently HMR hooks are not documented. If some day they are, then it must uses `HotModuleReplacementPlugin.hooks(parser)`.

### CODE REVIEW

**Comment by ooflorent** [comment]
@jantimon Maybe you want to look at this PR!

**Comment by jantimon** [comment]
@ooflorent thanks :) actually I really like the `WeakMap` approach.
It feels cleaner than modifying the webpack internal hooks directly.  
Unfortunately it would be a huge breaking change for my plugin.

**Comment by ooflorent** [comment]
For webpack also. That's why we change this for the next major (v5).

**Comment by jantimon** [comment]
I am a huge fan of semantic versioning however releasing a new major often breaks up with the current eco system.

For the new version of the html-webpack-plugin I had to change the hooks anyway so I'll also switch to a `WeakMap` and export the content of https://github.com/jantimon/html-webpack-plugin/blob/webpack-4/lib/hooks.js  as a static helper 👍

Thanks for sharing!! :)

File: lib/HotModuleReplacementPlugin.js [hunk 1]
```diff
@@ -18,7 +18,25 @@ const {
...
 } = require("./JavascriptParserHelpers");
 
+const hotInitCode = Template.getFunctionContent(
+	require("./HotModuleReplacement.runtime.js")
+);
+
+const parserHooksMap = new WeakMap();
+
 module.exports = class HotModuleReplacementPlugin {
+	static hooks(parser) {
```
Comment (sokra) [inline lib/HotModuleReplacementPlugin.js]: getHooks

**Comment by sokra** [comment]
Thanks

**Pull request merged**